broker-daemon/state-machines/interchain.js

```diff
@@ -0,0 +1,291 @@
+  } catch (e) {
+    if (e instanceof ENGINE_ERRORS.SettledSwapError) {
+      logger.debug(`Swap for ${hash} has already been settled`)
+
+      return inboundEngine.getSettledSwapPreimage(hash)
+    }
+
+    throw e
+  }
+
+  // add our static time lock to the time the inbound contract was locked
+  // in to arrive at the latest time that our outbound contract can be
+  // resolved while still considering our state "safe" and atomic.
+  const maxTime = new Date(committedTime.getTime() + (OUTBOUND_TIME_LOCK * 1000))
+
+  logger.debug(`Sending payment to ${outboundAddress} to translate ${hash}`, {
+    maxTime,
+    outboundAmount
+  })
+
+  return outboundEngine.translateSwap(
+    outboundAddress,
+    hash,
+    outboundAmount,
+    maxTime
+  )
+}
+
+/**
+ * Handle errors encountered while translating swaps across chains.
+ *
+ * @private
+ *
+ * @param   {Error} error                   - Error encountered to be handled
+ * @param   {string}  hash                  - Base64 string of the hash for the
+ *                                            swap
+ * @param   {Payment} inboundPayment        - Expected inbound payment
+ * @param   {Payment} outboundPayment       - Outbound payment we will make to
+ *                                            retrieve the preimage.
+ * @returns {string}                          Base64 encoded preimage for the swap
+ * @throws  {Error} If a permanent error is encountered and the swap is cancelled
+ */
+async function handleSwapError (error, hash, inboundPayment, outboundPayment) {
+  if (error instanceof ENGINE_ERRORS.PermanentSwapError) {
+    logger.error('Permanent Error encountered while translating swap, ' +
+      'cancelling upstream invoice', { error: error.message, hash })
+
+    await inboundPayment.engine.cancelSwap(hash)
+
+    // once we've cancelled, we can re-throw
+    throw error
+  }
+
+  logger.error('Temporary Error encountered while translating swap',
+    { error: error.stack, hash })
+
+  // A temporary error means we don't know the current state, so we need
+  // to restart the whole process
+  logger.debug(`Delaying swap retry for ${hash} for ${RETRY_DELAY}ms`)
+  await delay(RETRY_DELAY)
+  logger.debug(`Retrying swap translation for ${hash}`)
+
+  return translateSwap(hash, inboundPayment, outboundPayment)
+}
+
+/**
+ * Translate a swap cross-chain by retrieving the preimage from the
+ * downstream chain and returning it to the upstream chain.
+ *
+ * @private
+ *
+ * @param {string}  hash                  - Base64 string of the hash for the
+ *                                          swap
+ * @param {Payment} inboundPayment        - Expected inbound payment
+ * @param {Payment} outboundPayment       - Outbound payment we will make to
+ *                                          retrieve the preimage.
+ * @returns {string}                        Base64 encoded preimage for the swap
+ */
+async function translateOnce (hash, inboundPayment, outboundPayment) {
```

Comment:
Since `engine.translateSwap` occurs in `getPreimage`, it looks like we're using `translate` to mean `translateSwap` + `settleSwap`, we might want to use a different name like `executeSwap` for `translateSwap` + `settleSwap`. Also on naming: `getPreimage` sounds like it would only be getting an existing preimage rather than creating a new one with `translateSwap`. Perhaps we could rename `getPreimage` to something like `translateSwapWhenReady` and the "getPreimage" aspect is just the idempotence of the "translateSwap" operation.